File: C++/섹션08. A형 문제 풀이 (~2022 상반기)/02. 2개의 사탕 (코드트리, 2015 하반기 2번).cpp
#include <stdio.h>

#define MAX (10 + 5)
#define INF (0x7fff0000)

int T;

int N, M;
char MAP[MAX][MAX];

int num_of_cases[10 + 5];

struct RC
{
	int r;
	int c;
};

RC RED, BLUE;

int minAnswer;

// ↑, →, ↓, ←
int dr[] = { -1, 0, 1, 0 };
int dc[] = { 0, 1, 0, -1 };

void input()
{
	scanf("%d %d", &N, &M);

	for (int r = 0; r < N; r++)
	{
		for (int c = 0; c < M; c++)
		{
			scanf(" %c", &MAP[r][c]);

			if (MAP[r][c] == 'R')
			{
				RED.r = r;
				RED.c = c;
				
				MAP[r][c] = '.';
			}
			else if (MAP[r][c] == 'B')
			{
				BLUE.r = r;
				BLUE.c = c;

				MAP[r][c] = '.';
			}
		}
	}
}

void printMap() // for debug
{
	MAP[RED.r][RED.c] = 'R';
	MAP[BLUE.r][BLUE.c] = 'B';

	for (int r = 0; r < N; r++) 
		printf("%s\n", MAP[r]);
	putchar('\n');

	MAP[RED.r][RED.c] = '.';
	MAP[BLUE.r][BLUE.c] = '.';
}

void printCases()
{
	for (int i = 0; i < 10; i++) printf("%d ", num_of_cases[i]);
	putchar('\n');
}

int simulate()
{
	RC red, blue;

	red = RED;
	blue = BLUE;

	for (int i = 0; i < 10; i++)
	{
		int direction = num_of_cases[i];
	
		while (1)
		{
			int red_nr = red.r + dr[direction];
			int red_nc = red.c + dc[direction];
			int blue_nr = blue.r + dr[direction];
			int blue_nc = blue.c + dc[direction];

			bool checkWallRed = MAP[red_nr][red_nc] == '#';
			bool checkWallBlue = MAP[blue_nr][blue_nc] == '#';
			bool checkCandyRed = (red_nr == blue.r) && (red_nc == blue.c);
			bool checkCandyBlue = (blue_nr == red.r) && (blue_nc == red.c);

			// for debug
			// RED = red; BLUE = blue;
			// printMap();

			// 둘 다 다음 좌표가 벽이면 break;
			if (checkWallRed == true && checkWallBlue == true) break;

			// 둘 다 다음 좌표가 벽이 아니라면, 1칸 이동
			if (checkWallRed == false && checkWallBlue == false)
			{
				red.r = red_nr;
				red.c = red_nc;

				blue.r = blue_nr;
				blue.c = blue_nc;
			}
			// 둘 중 하나의 사탕만 움직이는 경우
			else if (checkWallRed == true && checkCandyBlue == false)
			{
				blue.r = blue_nr;
				blue.c = blue_nc;
			}
			else if (checkWallBlue == true && checkCandyRed == false)
			{
				red.r = red_nr;
				red.c = red_nc;
			}
			// 그 외의 모든 경우 break
			else
				break;

			if (MAP[red.r][red.c] == 'O')
			{
				while (MAP[blue.r][blue.c] != '#')
				{
					if (MAP[blue.r][blue.c] == 'O') return INF;

					blue.r += dr[direction];
					blue.c += dc[direction];
				}

				return i + 1;
			}

			if (MAP[blue.r][blue.c] == 'O') return INF;
		}
	}

	return INF;
}

void DFS(int depth, int direction)
{
	if (depth == 10)
	{
		// printCases();

		int tmp = simulate();
		if (tmp < minAnswer) minAnswer = tmp;
				
		return;
	}

	for (int i = 0; i < 4; i++)
	{
		// 연속된 방향 무시
		if (i == direction) continue;

		num_of_cases[depth] = i;

		DFS(depth + 1, i);
	}
}

int main()
{
	// scanf("%d", &T);
	T = 1;
	for (int tc = 1; tc <= T; tc++)
	{
		input();

		minAnswer = INF;
		
		DFS(0, -1);

		if (minAnswer == INF) printf("-1\n");
		else printf("%d\n", minAnswer);
	}
	
	return 0;
}

File: Python/섹션08. A형 문제 풀이 (~2022 상반기)/02. 2개의 사탕 (코드트리, 2015 하반기 2번).py
MAX = 10 + 5
INF = 0x7fff0000

MAP = [[0] * MAX for _ in range(MAX)]
num_of_cases = [0] * MAX

class RC:
    def __init__(self, r: int, c: int):
        self.r = r
        self.c = c

# ↑, →, ↓, ←
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

def input_data():
    global N, M, RED, BLUE
    
    N, M = map(int, input().split())
    for r in range(N):
        MAP[r] = list(input())
    
    for r in range(N):
        for c in range(M):
            if MAP[r][c] == 'R': 
                RED = RC(r, c)
                MAP[r][c] = '.'
            elif MAP[r][c] == 'B': 
                BLUE = RC(r, c)
                MAP[r][c] = '.'
         
def printMap(): # for debug
    MAP[RED.r][RED.c] = 'R'
    MAP[BLUE.r][BLUE.c] = 'B'
    
    for r in range(N): print("".join(MAP[r][:M]))
        
    MAP[RED.r][RED.c] = '.'
    MAP[BLUE.r][BLUE.c] = '.'    
    
def printCases():
    print(*num_of_cases[:10])
    
def simulate():
    red = RC(RED.r, RED.c)
    blue = RC(BLUE.r, BLUE.c)
    
    for i in range(10):
        direction = num_of_cases[i]
        
        while 1:
            red_nr = red.r + dr[direction]
            red_nc = red.c + dc[direction]
            blue_nr = blue.r + dr[direction]
            blue_nc = blue.c + dc[direction]
            
            checkWallRed = MAP[red_nr][red_nc] == '#'
            checkWallBlue = MAP[blue_nr][blue_nc] == '#'
            checkCandyRed = (red_nr == blue.r) and (red_nc == blue.c)
            checkCandyBlue = (blue_nr == red.r) and (blue_nc == red.c)
            
            # 둘 다 다음 좌표가 벽이면 break
            if checkWallRed == True and checkWallBlue == True: break
            
            # 둘 다 다음 좌표가 벽이 아니라면, 1칸 이동
            if checkWallRed == False and checkWallBlue == False:
                red.r = red_nr
                red.c = red_nc
                
                blue.r = blue_nr
                blue.c = blue_nc
            # 둘 중 하나의 사탕만 움직이는 경우
            elif checkWallRed == True and checkCandyBlue == False:
                blue.r = blue_nr
                blue.c = blue_nc
            elif checkWallBlue == True and checkCandyRed == False:
                red.r = red_nr
                red.c = red_nc
            # 그 외 모든 경우 break
            else:
                break
            
            if MAP[red.r][red.c] == 'O':
                while MAP[blue.r][blue.c] != '#':
                    if MAP[blue.r][blue.c] == 'O': return INF
                    
                    blue.r += dr[direction]
                    blue.c += dc[direction]
                    
                return i + 1
   
            if MAP[blue.r][blue.c] == 'O': return INF

    return INF
      
def DFS(depth, direction):
    global minAnswer
    
    if depth == 10:
        # printCases()
        
        tmp = simulate()
        if tmp < minAnswer: minAnswer = tmp
        
        return

    for i in range(4):
        # 연속된 방향 무시
        if i == direction: continue
        
        num_of_cases[depth] = i
        
        DFS(depth + 1, i)
                    
# T = int(input())
T = 1
for tc in range(T):
    global minAnswer
    
    input_data()
        
    minAnswer = INF
    
    DFS(0, -1)
    
    if minAnswer == INF: print(-1)
    else: print(minAnswer)